spec/ruby/core/module/name_spec.rb

```diff
@@ -6,6 +6,40 @@
     Module.new.name.should be_nil
   end
 
+  ruby_version_is "3.3" do
+    it "can assign a temporary name" do
+      m = Module.new
+      m.name.should be_nil
+
+      m.set_temporary_name("Foo")
+      m.name.should == "Foo"
```

Comment:
I'm uncomfortable with this, it means `Class#name` is completely disconnected from the constant path (i.e. it's unrelated to constant `Foo`).

Can we at least prevent a temporary name starting with an uppercase letter to be able to easily distinguish these "fake" names from real constant paths?

Reply:
What I think we should introduce later, is `Module#permanent?` or `Module#anonymous?` or some way you can check whether a name is actually a real constant or not. I'm also okay with updating `#to_s` and `#inspect` to reflect this if that's compatible enough.

I personally think your proposed change is too restrictive. One use case for setting a "proper" name on an anonymous module is to do something like this:

```
module Template
  def self.[](path)
    self.new(path)
  end

  def self.new(path)
    m = Class.new
    m.set_temporary_name("#{self}[#{path.dump}]")
    # ...
    return m
  end
end

Template["foo.rb"]
# => Template["foo.rb"]
```

Reply:
> `Module#permanent?` or `Module#anonymous?`

This wouldn't help anyone e.g. seeing an error message containing that fake constant name.
But it could be useful to tell if Class#name is related to a constant path or not.

> I'm also okay with updating #to_s and #inspect to reflect this if that's compatible enough.

This could work, if it also affects the name shown in exception messages (`rb_mod_name()` is used I think).
It should be compatible, since so far anonymous modules have anyway a anonymous-like `to_s/inspect`.

> I personally think your proposed change is too restrictive. One use case for setting a "proper" name on an anonymous module is to do something like this:

It could be `template["foo.rb"]` and then it is immediately clear to the reader there is no Template constant, unlike `undefined method 'foo' for #<Template["foo.rb"]:0x00007efc38711fc0> (NoMethodError)`.
But yeah it wouldn't work to actually eval `template["foo.rb"]`.

We could also check if the temporary name is a valid constant name, and only forbid those.
At least that name is a not a valid constant name:
```
irb(main):003:0> m=Module.new
=> #<Module:0x00007fa89c26a970>
irb(main):004:0> m.const_set('Template["foo.rb"]', 42)
(irb):4:in `const_set': wrong constant name Template["foo.rb"] (NameError)
```
It is harder to visually notice though.